### PR TITLE
Use envoy v2 bootstrap for sidecar

### PR DESCRIPTION
**Pull request by costinm**
Add generation of the sidecar bootstrap. 
Split from the larger PR dealing with SNI ingress, to unblock other people who need to test with v2.

### CODE REVIEW

File: install/kubernetes/addons/kube-lego.yaml [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        - name: LEGO_POD_IP
+          valueFrom:
+            fieldRef:
+              fieldPath: status.podIP
+        readinessProbe:
+          httpGet:
+            path: /healthz
+            port: 8080
+          initialDelaySeconds: 5
+          timeoutSeconds: 1
```
Comment (rshriram) [inline install/kubernetes/addons/kube-lego.yaml]: can you split this out into a separate PR?

File: install/kubernetes/ingress-sni-zvpn.yaml [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+      access_log_path: /dev/stdout
+      address:
+        socket_address: { address: 0.0.0.0, port_value: 15000 }
+
+    dynamic_resources:
+      lds_config:
+        api_config_source:
+          api_type: REST_LEGACY
+          refresh_delay: 10s
+          cluster_name: [rds]
```
Comment (rshriram) [inline install/kubernetes/ingress-sni-zvpn.yaml]: rds?

File: install/kubernetes/ingress-sni-zvpn.yaml [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+    admin:
+      access_log_path: /dev/stdout
+      address:
+        socket_address: { address: 0.0.0.0, port_value: 15000 }
+
+    dynamic_resources:
+      lds_config:
+        api_config_source:
+          api_type: REST_LEGACY
+          refresh_delay: 10s
```
Comment (rshriram) [inline install/kubernetes/ingress-sni-zvpn.yaml]: This sucks.. coz we have global mesh configs to control this. Is there any way we can push this into the pilot agent to spit out the YAML based on global configs?

Comment (rshriram) [reply install/kubernetes/ingress-sni-zvpn.yaml]: and why do you need LDS at all for ingress? You just need CDS, RDS and EDS

File: install/kubernetes/ingress-sni-zvpn.yaml [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+                #              - name: envoy.router
+                #                deprecated_v1: true
+
+      # HTTPS - currently manually configured, until pilot generates LDSv2 for ingress.
+      # If using kube-lego, certificate can be retrieved for testing with:
+      #   kubectl get secret -n istio-system httpbin-tls -o jsonpath='{.data.tls\.key}' |base64 -d > /etc/istio/httpbin-certs/tls.key
+      #   kubectl get secret -n istio-system httpbin-tls -o jsonpath='{.data.crt\.key}' |base64 -d > /etc/istio/httpbin-certs/crt.key
+      - name: http443
+        address:
+          socket_address: { address: 0.0.0.0, port_value: 8443 }
```
Comment (rshriram) [inline install/kubernetes/ingress-sni-zvpn.yaml]: wrong port 8443

File: install/kubernetes/ingress-sni-zvpn.yaml [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+              deprecated_v1: true
+              value:
+                stat_prefix: tcp_pilot
+                route_config:
+                    routes:
+                    - cluster: zvpn.pilot
+
+      - name: tcp-zvpn-ca
+        address:
+          socket_address: { address: 0.0.0.0, port_value: 8060 }
```
Comment (rshriram) [inline install/kubernetes/ingress-sni-zvpn.yaml]: why expose CA

Comment (rshriram) [reply install/kubernetes/ingress-sni-zvpn.yaml]: or pilot, etc. ?

File: install/kubernetes/addons/kube-lego.yaml (hunk 1 above)
Comment (costinm) [reply install/kubernetes/addons/kube-lego.yaml]: Sure, lego is needed so I can test the certs.

File: install/kubernetes/ingress-sni-zvpn.yaml (hunk 2 above)
Comment (costinm) [reply install/kubernetes/ingress-sni-zvpn.yaml]: Funny - current LDS v1 hardcodes the cluster name for discovery as rds

I could define 2 clusters, pointing to same pilot - but once we use LDSv2 it won't be a problem, and I can use [xds] name.

File: install/kubernetes/ingress-sni-zvpn.yaml (hunk 4 above)
Comment (costinm) [reply install/kubernetes/ingress-sni-zvpn.yaml]: Currently it's setup to have both LDS v1 (from pilot ) and those v2 listeners - which don't work yet completely. If I use port 443 in the manual one - it'll be replace by the LDSv1. 

This is temporary, to test and compare both. Once I get v2 working ( there are at least 2 problems with Oct Envoy that gets into proxy ) - I'll start working on generating the working format from pilot.

File: install/kubernetes/ingress-sni-zvpn.yaml (hunk 5 above)
Comment (costinm) [reply install/kubernetes/ingress-sni-zvpn.yaml]: That's part of the 'zero vpn' - machines without VPN/VPC in other clusters (or raw VMs) should be able to access pilot and CA (if they have bootstrap certs). 

It's a plain TCP proxy, with TLS terminated in CA and pilot.

Comment (costinm) [reply install/kubernetes/ingress-sni-zvpn.yaml]: To be clear - my primary goal with the PR is to get a working zero-vpn ( mesh expansion without requiring the internal load balancer and VPC/VPN). The SNI is a secondary goal, doing them at the same time since both require v2

File: install/kubernetes/addons/kube-lego.yaml (hunk 1 above)
Comment (costinm) [reply install/kubernetes/addons/kube-lego.yaml]: BTW - any preference of where to put it ? I don't think it should be installed by default with all the addons, and will need some customization ( probably need to remove my email ). Should I move it to samples ?

**Comment by danehans** [comment]
xref: https://github.com/istio/istio/pull/2561

File: install/kubernetes/ingress-sni-zvpn.yaml (hunk 3 above)
Comment (costinm) [reply install/kubernetes/ingress-sni-zvpn.yaml]: LDS is need for one of the zero-vpn solutions - so we can add ports on the ingress for mesh expansion.

Comment (costinm) [reply install/kubernetes/ingress-sni-zvpn.yaml]: Also for TCP ingress, when Gateway is added.

**Comment by costinm** [comment]
Waiting for envoy to be updated...

File: install/kubernetes/ingress-sni-zvpn.yaml (hunk 5 above)
Comment (wattli) [reply install/kubernetes/ingress-sni-zvpn.yaml]: So is this some prototype or real implementation?

Comment (costinm) [reply install/kubernetes/ingress-sni-zvpn.yaml]: It's the first step towards a real implementation :-) It should be usable and work - but there are many changes needed to improve the UX and simplify. Step one is getting SNI and zvpn using 
a manually configuration, step 2 is to automate..

**Comment by rshriram** [comment]
Is this still wip?

File: install/kubernetes/ingress-sni-zvpn.yaml [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+#
+# Setting up SNI in LDSv2:
+# https://www.envoyproxy.io/docs/envoy/latest/faq/sni.html#faq-how-to-setup-sni
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: envoy-config
+  namespace: istio-system
+data:
+  envoy.yaml: |
```
Comment (kyessenov) [inline install/kubernetes/ingress-sni-zvpn.yaml]: Now that we imported go-control-plane, can we use the schematize version of this YAML?
We want to preserve backwards compatibility with the existing switches in the mesh config (turning on and off mixer, tracing, etc).

File: install/kubernetes/ingress-sni-zvpn.yaml (hunk 3 above)
Comment (kyessenov) [reply install/kubernetes/ingress-sni-zvpn.yaml]: Please maintain compatibility with the mesh config. I don't see how to turn on/off tracing, for example, using this YAML.

File: install/kubernetes/addons/kube-lego.yaml (hunk 1 above)
Comment (kyessenov) [reply install/kubernetes/addons/kube-lego.yaml]: Please submit a separate PR. It's generally useful, and easier to manager as a feature (in case we need to roll back).

File: install/kubernetes/ingress-sni-zvpn.yaml [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+
+      - name: tcp-zvpn-pilot
+        address:
+          socket_address: { address: 0.0.0.0, port_value: 15003 }
+        filter_chains:
+        - filters:
+          - name: envoy.tcp_proxy
+            config:
+              # listener_manager_impl.cc - if false, will attempt to convert to proto - not implemented for tcp_proxy
+              deprecated_v1: true
```
Comment (kyessenov) [inline install/kubernetes/ingress-sni-zvpn.yaml]: See https://github.com/envoyproxy/envoy/issues/2441

File: install/kubernetes/ingress-sni-zvpn.yaml [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+      - name: zvpn.pilot
+        connect_timeout: 0.25s
+        type: STRICT_DNS
+        lb_policy: ROUND_ROBIN
+        hosts:
+          - { socket_address: { address: "istio-pilot", port_value: 15003 }}
+
+      # Connection to Istio CA. Service name must match the one for CA exposing gRPC.
+      - name: zvpn.ca
+        connect_timeout: 0.25s
```
Comment (kyessenov) [inline install/kubernetes/ingress-sni-zvpn.yaml]: No TLS, no protocol options?

File: pilot/docker/envoy_ingress.yaml [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+#
+# Current config has a sample port open, exposing an internal service. This will be rolled into LDS and automatically
+# exposed using service annotations in a future version.
+#
+# To customize the envoy config, create a config map and update the ingress config to use the custom envoy bootstrap
+# from the config map.
+#
+# Setting up SNI in LDSv2:
+# https://www.envoyproxy.io/docs/envoy/latest/faq/sni.html#faq-how-to-setup-sni
+stats_config:
```
Comment (kyessenov) [inline pilot/docker/envoy_ingress.yaml]: This is going to be very challenging to maintain given the amount of YAML present in this PR, and not quite stable API surface that it builds upon.

Please use templating or code to share as much as possible.

File: pilot/pkg/proxy/envoy/config.go [hunk 10]
```diff
@@ -78,6 +78,7 @@ func (conf *Config) Write(w io.Writer) error {
 
 // BuildConfig creates a proxy config with discovery services and admin port
 // it creates config for Ingress, Egress and Sidecar proxies
+// TODO: remove after new agent package is done
```
Comment (kyessenov) [inline pilot/pkg/proxy/envoy/config.go]: I suggest to keep it and remove the hard-coded YAML instead.

File: pkg/agent/config.go [hunk 11]
```diff
@@ -0,0 +1,135 @@
+// Copyright 2017 Istio Authors
```
Comment (kyessenov) [inline pkg/agent/config.go]: bad merge?

File: install/kubernetes/ingress-sni-zvpn.yaml (hunk 3 above)
Comment (rshriram) [reply install/kubernetes/ingress-sni-zvpn.yaml]: That’s a good catch kuat. Costin this config is going to disable tracing at Ingress effectively. One option is to always enable tracing and bear with Envoy complaining that there is no associated cluster.

File: install/kubernetes/ingress-sni-zvpn.yaml [hunk 12]
```diff
@@ -0,0 +1,364 @@
...
+      lds_config:
+        api_config_source:
+          api_type: REST_LEGACY
+          refresh_delay: 10s
+          # Pilot cluster is named 'rds' for compatibility with existing LDS, which hardcodes this name for the RDS API.
+          cluster_names: [rds]
+      cds_config:
+         api_config_source:
+            api_type: REST_LEGACY
+            refresh_delay: 10s
```
Comment (rshriram) [inline install/kubernetes/ingress-sni-zvpn.yaml]: Why don’t you set this to 1s - same as what we use today.

File: install/kubernetes/ingress-sni-zvpn.yaml [hunk 13]
```diff
@@ -0,0 +1,364 @@
...
+      # Example for testing explicit routes (route_config vs rds)
+      - name: http8081
+        address:
+          socket_address: { address: 0.0.0.0, port_value: 8081 }
+        filter_chains:
+        - filters:
+          - name: envoy.http_connection_manager
+            # TODO: add mixer
+            config:
+              stat_prefix: ingress_http8080
```
Comment (rshriram) [inline install/kubernetes/ingress-sni-zvpn.yaml]: 8081

File: install/kubernetes/ingress-sni-zvpn.yaml [hunk 14]
```diff
@@ -0,0 +1,364 @@
...
+              value:
+                stat_prefix: tcp_ca
+                route_config:
+                    routes:
+                    - cluster: zvpn.ca
+
+
+      clusters:
+      - name: rds
+        connect_timeout: 0.25s
```
Comment (rshriram) [inline install/kubernetes/ingress-sni-zvpn.yaml]: Also sync with mesh config. I think it was 10s

File: install/kubernetes/ingress-sni-zvpn.yaml [hunk 15]
```diff
@@ -0,0 +1,364 @@
...
+          # Local debugging/testing
+          #- { socket_address: { address: "localhost", port_value: 15008 }}
+
+      # Direct zvpn connection to pilot TLS port. This is used with a TCP proxy !
+      - name: zvpn.pilot
+        connect_timeout: 0.25s
+        type: STRICT_DNS
+        lb_policy: ROUND_ROBIN
+        hosts:
+          - { socket_address: { address: "istio-pilot", port_value: 15003 }}
```
Comment (rshriram) [inline install/kubernetes/ingress-sni-zvpn.yaml]: Name not properly qualified.

File: install/kubernetes/ingress-sni-zvpn.yaml (hunk 8 above)
Comment (rshriram) [reply install/kubernetes/ingress-sni-zvpn.yaml]: No protocol options as it’s a tcp cluster. No tls either since it’s passthrough tcp cluster

File: install/kubernetes/ingress-sni-zvpn.yaml [hunk 16]
```diff
@@ -0,0 +1,364 @@
...
+        connect_timeout: 0.25s
+        lb_policy: ROUND_ROBIN
+        type: EDS
+        eds_cluster_config:
+          service_name: istio-ca-ilb.istio-system.svc.cluster.local|grpc
+          eds_config:
+            api_config_source:
+              api_type: REST_LEGACY
+              cluster_names: [rds]
+              refresh_delay: 10s
```
Comment (rshriram) [inline install/kubernetes/ingress-sni-zvpn.yaml]: If you want to ship this, please set it to simple strict dns cluster.

File: install/kubernetes/ingress-sni-zvpn.yaml [hunk 17]
```diff
@@ -0,0 +1,364 @@
...
+        # Will auto-update on gcr.io pushes
+        keel.sh/policy: force
+        keel.sh/trigger: poll
+    spec:
+      serviceAccountName: istio-ingress-service-account
+      containers:
+      - name: istio-ingress
+        # Manually start envoy, without using pilot-agent, so we can override the config.
+        command: ["/usr/local/bin/envoy"]
+        image: gcr.io/istio-testing/proxy_debug:costin
```
Comment (rshriram) [inline install/kubernetes/ingress-sni-zvpn.yaml]: Why? Use lyft/Envoy

File: install/kubernetes/ingress-sni-zvpn.yaml (hunk 6 above)
Comment (costinm) [reply install/kubernetes/ingress-sni-zvpn.yaml]: Not sure what you mean by 'schematize version' - but this will be turned into a template, this is the test version to verify the config works.

File: install/kubernetes/ingress-sni-zvpn.yaml (hunk 8 above)
Comment (costinm) [reply install/kubernetes/ingress-sni-zvpn.yaml]: Yes, that's the point of zvpn - TLS is end to end, this is just TCP proxy. We may add some 
extra options ( customizable timeout, etc)

File: pilot/docker/envoy_ingress.yaml (hunk 9 above)
Comment (costinm) [reply pilot/docker/envoy_ingress.yaml]: Yes, this will be replaced by the template - the code to use template is in progress, until it is 
ready it's useful to test ( so I know what to put in the template )

File: pilot/pkg/proxy/envoy/config.go (hunk 10 above)
Comment (costinm) [reply pilot/pkg/proxy/envoy/config.go]: The hard coded yaml will be replaced with a template, this code is copied from old agent to 
keep it compiling, I'm working on removing most code here and use template only.

File: install/kubernetes/ingress-sni-zvpn.yaml (hunk 12 above)
Comment (costinm) [reply install/kubernetes/ingress-sni-zvpn.yaml]: This - as well as tracing, statsd - will be added with the templating option. 

Re. LDS on ingress - it is needed for zvpn, will create all the TCP proxies.

File: install/kubernetes/ingress-sni-zvpn.yaml (hunk 16 above)
Comment (costinm) [reply install/kubernetes/ingress-sni-zvpn.yaml]: You mean the rds cluster ? Sure, will do this.

File: install/kubernetes/ingress-sni-zvpn.yaml (hunk 17 above)
Comment (costinm) [reply install/kubernetes/ingress-sni-zvpn.yaml]: That's istio proxy - current build is flip-floping to the 3-month old envoy which doesn't work.
Will be obviously replaced with the template version. 

Also missing are the mixer filters...

File: install/kubernetes/ingress-sni-zvpn.yaml (hunk 3 above)
Comment (costinm) [reply install/kubernetes/ingress-sni-zvpn.yaml]: Well, I'll match whatever the current agent does - just didn't get time to test the v2 way of configuring that, and I'm trying to get the template working first... WIP...

**Comment by mandarjog** [comment]
@costinm why are we using pb.text encoding? Yaml is much more readable.

File: tools/deb/envoy_bootstrap_tmpl.json [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+        }
+      }
+    }
+  },
+  "static_resources": {
+    "clusters": [
+      {
+        "name": "rds",
+        "type": "STRICT_DNS",
+        "connect_timeout": {{ .connect_timeout }},
```
Comment (kyessenov) [inline tools/deb/envoy_bootstrap_tmpl.json]: you can use human-readable duration here, e.g. `"5s"`.

File: tools/deb/envoy_bootstrap_tmpl.json [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+        "lb_policy": "ROUND_ROBIN",
+{{ if eq .config.ControlPlaneAuthPolicy 1 }}
+      "tls_context": {
```
Comment (kyessenov) [inline tools/deb/envoy_bootstrap_tmpl.json]: is white-spacing OK here? seems like it's off.

File: tools/deb/envoy_bootstrap_tmpl.json [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+      }
+    }
+  },
+  "dynamic_resources": {
+    "lds_config": {
+      "api_config_source": {
+        "api_type": "REST_LEGACY",
+        "refresh_delay": {{ .refresh_delay }},
+        "cluster_names": [
+          "rds"
```
Comment (kyessenov) [inline tools/deb/envoy_bootstrap_tmpl.json]: why is LDS pointing to RDS cluster?

File: tools/deb/envoy_bootstrap_tmpl.json [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+            "private_key": {
+              "filename": "/etc/certs/key.pem"
+            }
+          },
+          "validation_context": {
+            "trusted_ca": {
+              "filename": "/etc/certs/root-cert.pem"
+            },
+            "verify_subject_alt_name": [
+              "spiffe://cluster.local/ns/istio-system/sa/istio-pilot-service-account"
```
Comment (kyessenov) [inline tools/deb/envoy_bootstrap_tmpl.json]: these are customizable I believe, cc @myidpt

File: pilot/pkg/proxy/envoy/v1/watcher.go [hunk 22]
```diff
@@ -255,6 +257,14 @@ func NewProxy(config meshconfig.ProxyConfig, node string, logLevel string) proxy
 	}
 }
 
+// NewV2Proxy creates an instance of the proxy using v2 bootstrap
+func NewV2Proxy(config meshconfig.ProxyConfig, node string, logLevel string) proxy.Proxy {
```
Comment (kyessenov) [inline pilot/pkg/proxy/envoy/v1/watcher.go]: I think putting v2 into v1 envoy directory is against the intention of @rshriram to have two config generation engines.

File: pkg/bootstrap/envoy_bootstrap.go [hunk 23]
```diff
@@ -0,0 +1,91 @@
+// Copyright 2018 Istio Authors
```
Comment (kyessenov) [inline pkg/bootstrap/envoy_bootstrap.go]: does this need a dedicated package? can you update owners file to assign ownership to pkg/bootstrap?

Comment (costinm) [reply pkg/bootstrap/envoy_bootstrap.go]: It will be shared with node-agent, doesn't need additional owners, same rule as all shared packages. I think bikeshedding the directory name doesn't help - I did think about this and I believe this is the best location. If we add nginx we can add another file to generate nginx bootstrap here.

File: pilot/pkg/proxy/envoy/v1/watcher.go (hunk 22 above)
Comment (costinm) [reply pilot/pkg/proxy/envoy/v1/watcher.go]: For now it's the simplest / cleanest solution - we can move later, but eventually the entire v1 directory, including this part, will go away. Once SDS is available we won't need any code here - and this will move to bootstrap.

File: tools/deb/envoy_bootstrap_tmpl.json (hunk 21 above)
Comment (costinm) [reply tools/deb/envoy_bootstrap_tmpl.json]: They seem hardcoded in current pilot.

File: tools/deb/envoy_bootstrap_tmpl.json (hunk 20 above)
Comment (costinm) [reply tools/deb/envoy_bootstrap_tmpl.json]: LDS v1 hardcodeds "rds" for clusters - in v2 we have a single cluster for all pilot connections, 
so until we move to LDS v2 we'll use the 'rds' name for the cluster. After that we can name it 'pilot' or whatever else.

File: tools/deb/envoy_bootstrap_tmpl.json (hunk 19 above)
Comment (costinm) [reply tools/deb/envoy_bootstrap_tmpl.json]: Generated file seems ok.

File: tools/deb/envoy_bootstrap_tmpl.json (hunk 18 above)
Comment (costinm) [reply tools/deb/envoy_bootstrap_tmpl.json]: Can - but using the actual protobuf seems better, avoids extra conversions.

File: pkg/bootstrap/envoy_bootstrap.go (hunk 23 above)
Comment (kyessenov) [reply pkg/bootstrap/envoy_bootstrap.go]: Ok. I don't have a strong opinion about package, just not loosing association with envoy.

File: tools/deb/envoy_bootstrap_tmpl.json (hunk 20 above)
Comment (kyessenov) [reply tools/deb/envoy_bootstrap_tmpl.json]: Looking at Looking at https://github.com/istio/istio/blob/master/pilot/pkg/proxy/envoy/v1/testdata/envoy-sidecar-auth.json.golden#L33 I see different clusters for xDS services.

File: tools/deb/envoy_bootstrap_tmpl.json (hunk 19 above)
Comment (kyessenov) [reply tools/deb/envoy_bootstrap_tmpl.json]: I think you can use one of gotmpl tricks like `-}}` to avoid whitespace skew.

File: pkg/bootstrap/envoy_bootstrap.go [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+	"text/template"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+)
+
+// Generate the envoy v2 bootstrap configuration, using template.
+const (
+	// EpochFileTemplate is a template for the root config JSON
+	EpochFileTemplate = "envoy-rev%d.yaml"
+	DefaultCfgDir     = "/var/lib/istio/envoy/envoy_bootstrap_tmpl.json"
```
Comment (kyessenov) [inline pkg/bootstrap/envoy_bootstrap.go]: This seems like it should be a parameter, at least to enable testing. Maybe embed the template into a code file to simplify deployment?

File: pkg/bootstrap/envoy_bootstrap.go [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+	t, err := template.New("bootstrap").Parse(string(cfgTmpl))
+	if err != nil {
+		return "", err
+	}
+
+	opts := map[string]interface{}{
+		"config": config,
+	}
+
+	// Simplify the template
```
Comment (ayj) [inline pkg/bootstrap/envoy_bootstrap.go]: You could also [template.FuncMap](https://golang.org/pkg/text/template/#FuncMap) to define custom formatting functions, e.g. https://play.golang.org/p/31vk0V7uXls

File: pkg/bootstrap/envoy_bootstrap_test.go [hunk 26]
```diff
@@ -0,0 +1,80 @@
...
+		return nil, err
+	}
+	cfg := &meshconfig.ProxyConfig{}
+	err = proto.UnmarshalText(string(content), cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// Exported from makefile or env
+	cfg.ConfigPath = out + "/bootstrap/" + base
```
Comment (ayj) [inline pkg/bootstrap/envoy_bootstrap_test.go]: In this test going to fail if I run `go test` without first running the makefile to export dependent envvars?

File: tests/e2e/framework/framework.go [hunk 27]
```diff
@@ -185,10 +185,7 @@ func (t *testCleanup) cleanup() {
 // Fetch and save cluster pod logs using kuebctl
 // Logs are uploaded during test tear down
 func (c *CommonConfig) saveLogs(r int) error {
-	if c.Cleanup.skipCleanup {
```
Comment (ldemailly) [inline tests/e2e/framework/framework.go]: Please add another way to skip that rather than delete the feature entirely - it makes the dev mode (where you repeatedly run 1 test) a lot slower if there is no way to skip that stage

File: tools/deb/istio.mk [hunk 28]
```diff
@@ -49,15 +50,19 @@ ISTIO_DEB_NAME ?= istio-sidecar
 # since we need configuration.
 # --iteration 1 adds a "-1" suffix to the version that didn't exist before
 ${ISTIO_OUT}/istio-sidecar.deb: | ${ISTIO_OUT}
+	$(MAKE) deb/fpm
```
Comment (mattdelco) [inline tools/deb/istio.mk]: I'm not clear on why this is recursively calling make (vs. being a dependency), nor why this & deb/build-in-docker are using a name that looks like a directory path.  The "deb/fpm" (or whatever it ends up being called) can be added to the .PHONY list.

How was "make deb" insufficient for the debian build (and thus provoking the addition of deb/fpm?).  If it's about trying to bypass the dependencies (to avoid a possible rebuild attempt of the go binaries), then I suppose a possible option is to have deb/build-in-docker pass a flag to make that's used to disable a line like:

$(eval ${ISTIO_OUT}/istio-sidecar.deb: $(ISTIO_OUT)/$(DEP))

(might need to put the two $(eval) lines into their own foreach loop to avoid getting too hard to read).

File: tools/istio-docker.mk [hunk 29]
```diff
@@ -80,6 +80,7 @@ docker.eurekamirror: $(ISTIO_DOCKER)/pilot-test-eurekamirror
 docker.pilot:        $(ISTIO_DOCKER)/pilot-discovery
 docker.proxy docker.proxy_debug: $(ISTIO_DOCKER)/pilot-agent
 $(foreach FILE,$(PROXY_JSON_FILES),$(eval docker.proxy docker.proxy_debug: $(ISTIO_DOCKER)/$(notdir $(FILE))))
+docker.proxy docker.proxy_debug: $(ISTIO_DOCKER)/envoy_bootstrap_tmpl.json
```
Comment (mattdelco) [inline tools/istio-docker.mk]: The $(ISTIO_DOCKER)/envoy_bootstrap_tmpl.json could be added to the end of the existing "docker.proxy docker.proxy_debug:" that's just two lines up.

File: tests/e2e/framework/framework.go (hunk 27 above)
Comment (costinm) [reply tests/e2e/framework/framework.go]: Log fetching and cleanup skipping are very different things. I did try running the tests with this - I don't see any significant difference, but if someone wants to add a 'skipLogs' flag I don't mind (but please let the default to on - it is important to get the logs so we can see any unexpected errors ).

File: tools/deb/envoy_bootstrap_tmpl.json (hunk 20 above)
Comment (costinm) [reply tools/deb/envoy_bootstrap_tmpl.json]: Sure - v1 used multiple clusters, in v2 we use one.

File: tools/deb/istio.mk (hunk 28 above)
Comment (costinm) [reply tools/deb/istio.mk]: I'm trying to avoid and undo the evals - I think they are hard to read, I would rather have an
explicit clear rule.

In particular the 'build in docker' stuff doesn't seem to work anymore - all the magic with git seems to have affected it, will need to fix that as well.

File: tools/istio-docker.mk (hunk 29 above)
Comment (costinm) [reply tools/istio-docker.mk]: Same - I'm trying to have more readable makefile, and I'm planning to refactor the dockerfile creation (moving few files around) so it's more obvious what is going on.

File: pkg/bootstrap/bootstrap_config.go [hunk 30]
```diff
@@ -0,0 +1,94 @@
+// Copyright 2018 Istio Authors
```
Comment (rshriram) [inline pkg/bootstrap/bootstrap_config.go]: I would suggest having pkg/envoy/bootstrap.go

File: pkg/bootstrap/testdata/all.proto [hunk 31]
```diff
@@ -0,0 +1,16 @@
...
+service_cluster:         "istio-proxy"
+drain_duration:          {seconds: 5}
+parent_shutdown_duration: {seconds: 6}
+discovery_address:       "mypilot:1001"
+discovery_refresh_delay:  {seconds: 3}
+zipkin_address:           "localhost:6000"
+connect_timeout:          {seconds: 7}
+statsd_udp_address:       "10.1.1.1:9125"
+proxy_admin_port:         15003
+availability_zone:        "AZ"
```
Comment (rshriram) [inline pkg/bootstrap/testdata/all.proto]: what is this setting?

File: pkg/bootstrap/testdata/all_golden.json [hunk 32]
```diff
@@ -0,0 +1,121 @@
+{
```
Comment (rshriram) [inline pkg/bootstrap/testdata/all_golden.json]: nit: please name the file based on the config used to generate it.. bootstrap-with-auth.json.golden

**Comment by costinm** [comment]
I can rename the golden files in a separate PR if you don't mind - this has been going on for too long...

Re. package name - there is no envoy specific code, even if we're not planning other proxies. I'm trying to move away from the pilot organization, where agent code is mixed with server code.

Also I think we agreed that even LDS/etc are considered as generic APIs anyone can implement and use, so even in pilot it would make sense to stop using 'envoy' unless it's something very specific to
envoy.

File: tests/e2e/framework/framework.go (hunk 27 above)
Comment (ldemailly) [reply tests/e2e/framework/framework.go]: until recently none of the e2e ran with skip_cleanup which was a dev only shortcut so it'd be nice to preserve the dev feature (short dev+test cycles) and add some new stuff for ci as needed (and yes I agree it could be 2 flags instead of 1, but as long as we don't lose a way to get into that state)

**Comment by costinm** [comment]
If you feel useful to skip getting the logs - I don't mind a new flag.
My experience is that if logs are well organized, more is better. Android
'bugreport' system dump is huge and
has everything, and many times the problem is in an unexpected place.
If a test is passing - most developers will just ignore it, and we can skip
getting logs for passing tests.
Bug if a test fails - we should probably get more info - kubernetes events,
the actual config that
got applied, etc.

I'll look a bit if we can find a common way to get logs to files (and use
them in the 3-4 frameworks we have),
it may be something that can be used outside the test system.
